File: cpp/389_找不同_medium.cpp
class Solution {
public:
    char findTheDifference(string s, string t) {
        unordered_map<char, int> wordMap;
        for(char _s : s)
            if(wordMap.find(_s) == wordMap.end())
                wordMap[_s] = 1;
            else
                wordMap[_s]++;
        for(char _t : t)
            if(wordMap.find(_t) == wordMap.end())
                return _t;
            else{
                wordMap[_t]--;
                if(wordMap[_t]<0)
                    return _t;
            }
        return ' ';
    }
};

// 法2: 
class Solution {
public:
    char findTheDifference(string s, string t) {
        int res = 0;
        for(char _s : s)
            res ^= _s;
        for(char _t : t)
            res ^= _t;
        return char(res);
    }
};

File: python/389_找不同_medium.py
class Solution:
    def findTheDifference(self, s: str, t: str) -> str:
        wordMap = dict()
        for _s in s:
            if _s not in wordMap:
                wordMap[_s] = 1
            else:
                wordMap[_s] += 1
        for _t in t:
            if _t not in wordMap:
                return _t
            else:
                wordMap[_t] -= 1
                if wordMap[_t] < 0:
                    return _t
        return ''

### 法二：最终答案一定为奇数的那个，因此可以使用位异或运算
class Solution:
    def findTheDifference(self, s: str, t: str) -> str:
        res = 0
        for _s in s:
            res ^= ord(_s)
        for _t in t:
            res ^= ord(_t)
        return chr(res)